### PR TITLE
[3.x] Make `TextureButton` and `Button` update on texture change

### PR DESCRIPTION
**Pull request by AThousandShips**
* 3.x version of: #77159

Currently unable to verify functionality on 3.x builds, code should work but would appreciate someone verifying

### CODE REVIEW

**Review by YuriSizov — APPROVED** [review]
Logic slightly changes, as in 3.x we didn't check `if (destination == p_texture)` for TextureButton textures before. So with this change there won't be updates triggered if someone sets the exactly the same value to a member.

I think it should be fine though. And the code adaptation to 3.x LGTM.

File: scene/gui/button.cpp [hunk 1]
```diff
@@ -309,16 +310,28 @@ void Button::set_icon(const Ref<Texture> &p_icon) {
...
 Ref<Texture> Button::get_icon() const {
 	return icon;
 }
 
+void Button::_texture_changed() {
+	if (icon.is_valid()) {
+		update();
+		minimum_size_changed();
+	}
+}
```
Comment (YuriSizov) [inline scene/gui/button.cpp:333]: So one note here is that this is now what's called in `set_icon`, whereas in the 4.x version we just call the same methods unconditionally. It's a bit of a strange change between the two versions, but also I'm not sure why the check is here to begin with, in both 3.x and 4.x versions. In `TextureButton` we don't add a condition and it makes sense that we don't.

Comment (AThousandShips) [reply scene/gui/button.cpp:333]: The 4.x version does the check, taken directly from there

Comment (akien-mga) [reply scene/gui/button.cpp:333]: To add to this, the main risk I envision is no update happen when going from valid icon to no icon. (Currently compiling to check if it's indeed a problem.)

Comment (AThousandShips) [reply scene/gui/button.cpp:334]: Restored the order in the `set_icon` method

**Comment by AThousandShips** [comment]
Can remove the check for same value as well if wanted to keep parity with 3.x, though the changes to `TextureRect` added such a check in #81100

File: scene/gui/button.cpp (hunk 1 above)
Comment (AThousandShips) [reply scene/gui/button.cpp:334]: I don't really recall why I added that check in 4.x, can remove it here and in 4.x too

Comment (timothyqiu) [reply scene/gui/button.cpp:334]: I think the check is redundant because currently the only caller is the connected `changed` signal above.

Comment (AThousandShips) [reply scene/gui/button.cpp:334]: Agreed, the one in `TextureRect` looks like a leftover from earlier versions, leaving it in for the moment for parity

File: scene/gui/texture_button.cpp [hunk 2]
```diff
@@ -343,6 +344,27 @@ void TextureButton::set_focused_texture(const Ref<Texture> &p_focused) {
 	focused = p_focused;
 };
 
+void TextureButton::_set_texture(Ref<Texture> *p_destination, const Ref<Texture> &p_texture) {
+	DEV_ASSERT(p_destination);
```
Comment (timothyqiu) [inline scene/gui/texture_button.cpp:348]: `DEV_ASSERT` should only be used for bottleneck code. A regular check here won't matter.

But I think a better way is to pass by non-const reference, so we don't have to do any check.

Comment (AThousandShips) [reply scene/gui/texture_button.cpp:348]: See: https://github.com/godotengine/godot/pull/77159#discussion_r1209237736

Comment (timothyqiu) [reply scene/gui/texture_button.cpp:348]: > That won't work, assigning to the reference won't update the texture itself

The referenced `Ref` object won't change. What will be updated is the content of that `Ref`, which is the texture.

> Also `DEV_ASSERT` is for crashing, not performance critical AFAIK

A note was added recently:
https://github.com/godotengine/godot/blob/8449592d92eaeef990f5502b419d491ee3eeb7a6/core/error/error_macros.h#L797-L801

But anyway, I think it's fine if the Godot 4 version uses this. I should have read the original PR, sorry :(

**Review by timothyqiu — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by AThousandShips** [comment]
Thank you!

**Comment by akien-mga** [comment]
Cherry-picked for 3.5.3.

**Pull request merged**